module/interface.py

```diff
@@ -274,18 +314,180 @@ def codegen(self, folder, project_type='', parameters='vectors',
         cg.codegen(work, folder, python_ext_name, project_type,
                    embedded, force_rewrite, float_flag, long_flag)
 
-
-def solve(P=None, q=None, A=None, l=None, u=None, **settings):
+    def adjoint_derivative(self, dx=None, dy_u=None, dy_l=None,
+            P_idx=None, A_idx=None):
         """
-        Solve problem of the form 
-
-        minimize     1/2 x' * P * x + q' * x
-        subject to   l <= A * x <= u
-
-        solver settings can be specified as additional keyword arguments. 
-        This function disables the GIL because it internally performs 
-        setup solve and cleanup.
+        Compute adjoint derivative after solve.
         """
 
-        unpacked_data, settings = utils.prepare_data(P, q, A, l, u, **settings)
-        return _osqp.solve(*unpacked_data, **settings)
+        P, q = self._derivative_cache['P'], self._derivative_cache['q']
+        A = self._derivative_cache['A']
+        l, u = self._derivative_cache['l'], self._derivative_cache['u']
+
+        try:
+            results = self._derivative_cache['results']
+        except KeyError:
+            raise ValueError("Problem has not been solved. You cannot take derivatives. Please call the solve function.")
+
+        if results.info.status != "solved":
+            raise ValueError("Problem has not been solved to optimality. You cannot take derivatives")
+
+        m, n = A.shape
+        x = results.x
+        y = results.y
+        z = A.dot(x)
+        y_u = np.maximum(y, 0)
+        y_l = -np.minimum(y, 0)
+
+        if A_idx is None:
+            A_idx = A.nonzero()
+
+        if P_idx is None:
+            P_idx = P.nonzero()
+
+        if dy_u is None:
+            dy_u = np.zeros(m)
+        if dy_l is None:
+            dy_l = np.zeros(m)
+
+        # Make sure M matrix exists
+        if 'M' not in self._derivative_cache:
+            # Multiply second-third column by diag(y_u) and diag(y_l)
+            # to make the matrix symmetric
+            M = spa.bmat([
+                [P, A.T @ spa.diags(y_u), -A.T @ spa.diags(y_l)],
+                [spa.diags(y_u) @ A, spa.diags(A @ x - u), None],
+                [-spa.diags(y_l) @ A, None, spa.diags(l - A @ x)]
+            ]).tocsc()
+            self._derivative_cache['M'] = M
+
+        # Prepare rhs
+        d_sol = np.concatenate([dx, dy_u, dy_l])
+
+        # Normalized version to compensate for diag(y_u) and diag(y_l) above
+        d_sol = np.concatenate([dx,
+                                spa.diags(y_u) @ dy_u,
+                                spa.diags(y_l) @ dy_l])
+
+        r_sol = - qdldl.Solver(self._derivative_cache['M']).solve(d_sol)
```

Comment:
Would it ever be helpful to cache `M`'s factorization too?

Reply:
I think we no longer need this since we might cache it inside the solver object if we want to solve it for a bunch of different rhs.